Add demo for FileAttachmentForm & WebDav public components

labkey-bpatel opened the pull request:
#### Rationale
As part of WNPRC Purchasing module development, there was a need to make FileAttachmentForm & WebDav components public. So adding demo for these components as part of making them public.

#### Changes
- File attachment upload example that uses FileAttachmentForm & WebDav's uploadWebDavFile api.
- Panel to view uploaded attachments that uses WebDav's getWebDavFiles api.

labkey-bpatel committed:
Add file attachment upload example.

labkey-bpatel committed:
Add uploaded files display example. Update README.

labkey-bpatel committed:
Add dev urls.

labkey-bpatel committed:
Update imports

cnathe commented on demo/README.md:
```diff
@@ -11,6 +11,8 @@ There are three example React pages:
 - http://localhost:8080/labkey/home/demo-helloWorld.view?
 - http://localhost:8080/labkey/home/demo-queryModel.view?
 - http://localhost:8080/labkey/home/demo-todoList.view?
+- http://localhost:8080/labkey/home/demo-fileAttachmentForm.view?
```
above this list of URLs it says "there are three example React pages", we should update that to say "The Demo module includes the following example React pages:"

cnathe commented on demo/src/client/entryPoints.js:
```diff
@@ -25,5 +25,15 @@ module.exports = {
+        name: 'fileAttachmentForm',
+        title: 'File Attachment Example Page',
+        permission: 'insert',
+        path: './src/client/FileAttachmentPage'
```
I like how you kept all of the React for these two new entryPoints in a single client/FileAttachmentPage dir. To make it even more clear what goes with what, I might suggest adding a subdir for this case like the `displayFiles` dir and move the app.tsx and dev.tsx into it (something like `insertFiles`).

cnathe commented on demo/src/client/FileAttachmentPage/models.ts:
```diff
@@ -0,0 +1,35 @@
+
+export class SavedFileModel {
+    /**
+     * @hidden
```
curious what this is for and if needed, why it isn't on the model above?

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+        <>
+            <Panel
+                className='panel panel-default'
+                expanded={true}
+                onToggle={function () {
+                }}
```
if this isn't an expand/collapse panel, then I think you can just remove these.

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+                    {
+                        !fileAttachmentModel.savedFiles &&
+                        <div>
+                            No files to display. Please upload files by going <a href={ActionURL.getContextPath() + '/home/demo-fileAttachmentForm.view#'}>here</a>.
```
Looks like the path for the "here" link is hardcoded to go to the `home` project. is that intentional? I assume we'd just want to say on the current container.

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+                        <>
+                            {
+                                fileAttachmentModel?.savedFiles?.map((savedFile, idx) => {
+                                    return <div>
```
looks like the browser is giving a JS error/warning about the items of a map not having distinct "key" values. I think that would be something like:
```suggestion
                                    return <div key={savedFile.fileName}>
```

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+                            {
+                                fileAttachmentModel?.savedFiles?.map((savedFile, idx) => {
+                                    return <div>
+                                        <a href="#" onClick={showImg}>{savedFile.fileName}</a>
```
was there a reason you went with the `onClick` here instead of just using the href from the savedFile directly?
```suggestion
                                        <a href={savedFile.href} target="_blank">{savedFile.fileName}</a>
```

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+        });
+    }, []);
+
+    const showImg = useCallback((evt: any) => {
```
See comment below. I don't think this is needed if you just use the `savedFile.href` directly in the `<a>` tag

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+
+export const App : FC = memo(() => {
+
+    const [fileAttachmentModel, setFileAttachmentModel] = useState<FileAttachmentModel>(new FileAttachmentModel());
```
giving this state variable a default value like `new FileAttachmentModel()` causes the page to flash the "no files to display" text while the `getUploadedFiles` ajax call is happening. I'd suggest removing this default and leaving it undefined and then add in a case before your return like:

```
if (! fileAttachmentModel) {
    return <LoadingSpinner/>
}
```

cnathe commented on demo/src/client/FileAttachmentPage/FileAttachmentForm.tsx:
```diff
@@ -0,0 +1,63 @@
+        uploadMyAttachmentsToServer(fileAttachmentModel)
+            .then(r => {
+                if (r) {
+                    //Routes to home page. Users can view files uploaded via Files webpart.
+                    window.location.href = ActionURL.buildURL('project', 'begin', getServerContext().container.path)
```
What do you think of combining these two entryPoints / panels so that this page shows the upload form and then in a panel below it shows the `<FileDisplayPanel>`? Then after the user clicks the save button we can just do a `window.location.reload()` to refresh the page so that the user sees the connection between the two panels more easily?

If we don't make that change, then I would suggest at least changing this redirect so that it takes the user to the `filecontent-begin` action instead of `project-begin`.

cnathe commented on demo/src/client/FileAttachmentPage/FileDisplayPanel.tsx:
```diff
@@ -0,0 +1,71 @@
+                                    return <div>
+                                        <a href="#" onClick={showImg}>{savedFile.fileName}</a>
+                                    </div>
+                                })
```
I'd suggest adding in a button or link to allow the user to "go to the files" so that they can delete them or clean them up if needed (since this is a working page that will actually upload files to the server). You can sent the user to `ActionURL.buildURL('filecontent', 'begin')`

cnathe commented on demo/src/client/FileAttachmentPage/FileAttachmentForm.tsx:
```diff
@@ -0,0 +1,63 @@
+
+    const [fileAttachmentModel, setFileAttachmentModel] = useState<FileAttachmentModel>(new FileAttachmentModel());
+
+    //equivalent to componentDidMount and componentDidUpdate
+    useEffect(() => {
+
+    }, []);
```
remove?

cnathe commented on demo/src/client/FileAttachmentPage/FileAttachmentForm.tsx:
```diff
@@ -0,0 +1,63 @@
+
+    const fileAttachmentChange = useCallback((model:FileAttachmentModel)=> {
+        setFileAttachmentModel(model);
+    }, [fileAttachmentModel]);
```
this useCallback dependency array looks wrong. Should it be `[setFileAttachmentModel]` instead as that is the dependency it is using isntead of the `fileAttachmentModel`?

cnathe commented on demo/src/client/FileAttachmentPage/FileAttachmentForm.tsx:
```diff
@@ -0,0 +1,63 @@
+
+    return (
+        <>
+            <FileAttachmentPanel onInputChange={fileAttachmentChange} model={fileAttachmentModel}/>
```
Since this is a working page that will actually save files to the LK file system, I might suggest adding an info alert to that affect above this panel:
```
<Alert bsStyle="info">Note: saving uploaded files will place them in the LabKey filesystem for this container.</Alert>
```

cnathe commented on demo/src/client/FileAttachmentPage/FileAttachmentPanel.tsx:
```diff
@@ -0,0 +1,48 @@
+    return (
+            <Panel
+                className='panel panel-default'
+                expanded={true}
+                onToggle={function () {
+                }}
```
I think these can be removed.

cnathe commented on demo/src/client/FileAttachmentPage/actions.ts:
```diff
@@ -0,0 +1,57 @@
+import { MY_ATTACHMENTS_DIR } from "./constants";
+
+//Uploads files to the server using WebDav api.
+export async function uploadMyAttachmentsToServer(model: FileAttachmentModel): Promise<any> {
```
the function is `async` but I don't see an `await` in it

cnathe replied:
Also the return type has `any` in it, should that be `Promise<string[]>`?

cnathe commented on demo/src/client/FileAttachmentPage/actions.ts:
```diff
@@ -0,0 +1,57 @@
+
+            // Note: you can return other properties of the WebDavFile, below only returns file name & href
+            const displayFiles = webDavFilesResponse.get('files').valueSeq().map((file: WebDavFile) => {
+                return {fileName: file.name, href: file.href};
```
should this be creating and returning a `SavedFileModel`?

cnathe committed:
Merge branch 'develop' into fb_file_attachment_demo

# Conflicts:
#	demo/package-lock.json
#	demo/package.json

cnathe committed:
Updates per code review feedback
- collapse two entrypoints down to one so that the "upload files" panel and "view files" panel are on the same page
- other misc code cleanup
